third_party/ibis/ibis_postgres/client.py

```diff
@@ -12,98 +12,37 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from __future__ import annotations
-import parsy
-import re
-import ast
-import toolz
 import ibis.expr.datatypes as dt
 import ibis.expr.schema as sch
 from ibis import util
-from ibis.backends.postgres.client import PostgreSQLClient
 
+import sqlalchemy as sa
+from ibis.backends.postgres import Backend as PostgresBackend
+from ibis.backends.postgres.datatypes import _BRACKETS, _parse_numeric, _type_mapping
 
-def _get_schema_using_query(self, query: str) -> sch.Schema:
-    raw_name = util.guid()
-    name = self.con.dialect.identifier_preparer.quote_identifier(raw_name)
-    type_info_sql = f"""\
-SELECT
-attname,
-format_type(atttypid, atttypmod) AS type
-FROM pg_attribute
-WHERE attrelid = {raw_name!r}::regclass
-AND attnum > 0
-AND NOT attisdropped
-ORDER BY attnum
-"""
-    with self.con.connect() as con:
-        con.execute(f"CREATE TEMPORARY VIEW {name} AS {query}")
-        try:
-            type_info = con.execute(type_info_sql).fetchall()
-        finally:
-            con.execute(f"DROP VIEW {name}")
-    tuples = [(col, self._get_type(typestr)) for col, typestr in type_info]
-    return sch.Schema.from_tuples(tuples)
-
-
-_BRACKETS = "[]"
-_STRING_REGEX = (
-    """('[^\n'\\\\]*(?:\\\\.[^\n'\\\\]*)*'|"[^\n"\\\\"]*(?:\\\\.[^\n"\\\\]*)*")"""
-)
-
-
-def spaceless(parser):
-    return SPACES.then(parser).skip(SPACES)
-
-
-def spaceless_string(*strings: str):
-    return spaceless(
-        parsy.alt(*(parsy.string(s, transform=str.lower) for s in strings))
-    )
-
-
-SPACES = parsy.regex(r"\s*", re.MULTILINE)
-RAW_NUMBER = parsy.decimal_digit.at_least(1).concat()
-SINGLE_DIGIT = parsy.decimal_digit
-PRECISION = SCALE = NUMBER = RAW_NUMBER.map(int)
-
-LPAREN = spaceless_string("(")
-RPAREN = spaceless_string(")")
-
-LBRACKET = spaceless_string("[")
-RBRACKET = spaceless_string("]")
 
-LANGLE = spaceless_string("<")
-RANGLE = spaceless_string(">")
-
-COMMA = spaceless_string(",")
-COLON = spaceless_string(":")
-SEMICOLON = spaceless_string(";")
-
-RAW_STRING = parsy.regex(_STRING_REGEX).map(ast.literal_eval)
-FIELD = parsy.regex("[a-zA-Z_][a-zA-Z_0-9]*")
-
-
-def _parse_numeric(
-    text: str, ddp: tuple[int | None, int | None] = (None, None)
-) -> dt.DataType:
-    decimal = spaceless_string("decimal", "numeric").then(
-        parsy.seq(LPAREN.then(PRECISION.skip(COMMA)), SCALE.skip(RPAREN))
-        .optional(ddp)
-        .combine(dt.Decimal)
-    )
-
-    brackets = spaceless(LBRACKET).then(spaceless(RBRACKET))
-
-    pg_array = parsy.seq(decimal, brackets.at_least(1).map(len)).combine(
-        lambda value_type, n: toolz.nth(n, toolz.iterate(dt.Array, value_type))
-    )
-
-    ty = pg_array | decimal
-    return ty.parse(text)
+def _metadata(self, query: str) -> sch.Schema:
```

Comment:
What's the purpose of this _metadata method? What is it extending from the base client?

Reply:
The metadata method is mainly used for custom query validations for our purpose. It takes the custom query and returns the schema associated with the results of the query.